File: Day364_2025/Q364_GFG.java
class Node {
    int data;
    Node next;
    
    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class Q364_GFG {

    private Node reverseList(Node head) {
        Node prev = null;
        Node curr = head;
        Node next;

        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }

    public Node addTwoLists(Node head1, Node head2) {
        // reverse both lists
        Node rNum1 = reverseList(head1);
        Node rNum2 = reverseList(head2);

        int carry = 0;
        Node dummy = new Node(-1);
        Node tail = dummy;

        while (rNum1 != null || rNum2 != null || carry != 0) {
            int op1 = (rNum1 != null) ? rNum1.data : 0;
            int op2 = (rNum2 != null) ? rNum2.data : 0;

            int sum = op1 + op2 + carry;
            carry = sum / 10;
            int digit = sum % 10;

            tail.next = new Node(digit);
            tail = tail.next;

            if (rNum1 != null) rNum1 = rNum1.next;
            if (rNum2 != null) rNum2 = rNum2.next;
        }

        // reverse result
        Node head = reverseList(dummy.next);

        // remove leading zeros
        while (head != null && head.data == 0) {
            head = head.next;
        }

        return head;
    }
}

File: Day364_2025/Q364_GFG.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Solution:
    def reverseList(self, head):
        prev = None
        curr = head

        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt

        return prev

    def addTwoLists(self, head1, head2):
        rNum1 = self.reverseList(head1)
        rNum2 = self.reverseList(head2)

        carry = 0
        dummy = Node(-1)
        tail = dummy

        while rNum1 or rNum2 or carry:
            op1 = rNum1.data if rNum1 else 0
            op2 = rNum2.data if rNum2 else 0

            total = op1 + op2 + carry
            carry = total // 10
            digit = total % 10

            tail.next = Node(digit)
            tail = tail.next

            if rNum1:
                rNum1 = rNum1.next
            if rNum2:
                rNum2 = rNum2.next

        head = self.reverseList(dummy.next)

        # remove leading zeros
        while head and head.data == 0:
            head = head.next

        return head
